app/services/hyrax/workflow/sipity_actions_generator.rb

```diff
@@ -38,11 +39,39 @@ def call
       private
 
         def generate_state_diagram!
+          # look for old states that need to be removed and validate the states can be removed.
+          removed_states = validate_states_to_be_removed
+
+          # remove any actions that will no longer be needed by the workflow
+          remove_unused_actions
+
           actions_configuration.each do |configuration|
             Array.wrap(configuration.fetch(:name)).each do |name|
               StateMachineGenerator.generate_from_schema(workflow: workflow, name: name, **configuration.except(:name))
             end
           end
+
+          # remove any states that are no longer needed by the workflow
+          #  Note I am doing this after the update so nothing is still linking to the
+          #       state to keep it from being deleted.
+          removed_states.each(&:delete)
+        end
+
+        def validate_states_to_be_removed
+          new_states = actions_configuration.map { |a| a[:transition_to] }
+          removed_states = workflow.workflow_states.reject { |state| new_states.include?(state.name) }
+          removed_states.each do |state|
```

Comment:
Won't this only raise on the first removed state? Perhaps try something like this:

```ruby
removed_states_with_entities = removed_states.select { |state| state.entities.any? }
raise InvalidStateRemovalException.new("Workflow states cannot be removed when they contain entities: #{removed_states_with_entities.to_sentence}") if removed_states_with_entities.any?
```